File: NodeJS/sitemap_scraper_hasdata.js
const axios = require("axios");
const fs = require("fs");
const xml2js = require("xml2js");

const apiKey = "YOUR-API-KEY";
const sitemapUrl = "https://demo.nopcommerce.com/sitemap.xml";

const url = "https://api.hasdata.com/scrape/web";

const payload = {
  url: sitemapUrl,
  proxyType: "datacenter",
  proxyCountry: "US"
};

const headers = {
  "Content-Type": "application/json",
  "x-api-key": apiKey
};

(async () => {
  try {
    const response = await axios.post(url, payload, { headers });
    const xmlContent = response.data.content;

    const parser = new xml2js.Parser();
    const result = await parser.parseStringPromise(xmlContent);

    const urls = [];
    const urlEntries = result.urlset?.url || [];

    for (const entry of urlEntries) {
      if (entry.loc && entry.loc[0]) {
        urls.push(entry.loc[0]);
      }
    }

    fs.writeFileSync("output.json", urls.join("\n"), "utf-8");
    console.log("Links saved to output.json");
  } catch (error) {
    console.error("Error:", error.message);
  }
})();

File: Python/sitemap_scraper_hasdata.py
import requests
import json
import xml.etree.ElementTree as ET

api_key = "YOUR-API-KEY"
sitemap_url = "https://demo.nopcommerce.com/sitemap.xml"

url = "https://api.hasdata.com/scrape/web"

payload = json.dumps({
  "url": sitemap_url,
  "proxyType": "datacenter",
  "proxyCountry": "US"
})
headers = {
  'Content-Type': 'application/json',
  'x-api-key': api_key
}

response = requests.request("POST", url, headers=headers, data=payload)

root = ET.fromstring(response.json().get("content"))
namespace = {'ns': 'http://www.sitemaps.org/schemas/sitemap/0.9'}
links = [loc.text for loc in root.findall('.//ns:loc', namespace)]

with open("output.json", 'w', encoding='utf-8') as f:
    for link in links:
        f.write(link + '\n')


